File: packages/wasmvm/wasmlib/src/wasmtypes/scaddress.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub const SC_ADDRESS_ALIAS: u8 = 8;
pub const SC_ADDRESS_ED25519: u8 = 0;
pub const SC_ADDRESS_NFT: u8 = 16;
pub const SC_ADDRESS_ETH: u8 = 32;

pub const SC_LENGTH_ALIAS: usize = 33;
pub const SC_LENGTH_ED25519: usize = 33;
pub const SC_LENGTH_NFT: usize = 33;
pub const SC_LENGTH_ETH: usize = 20;

pub const SC_ADDRESS_LENGTH: usize = SC_LENGTH_ED25519;

#[derive(PartialEq, Clone)]
pub struct ScAddress {
    pub(crate) id: [u8; SC_ADDRESS_LENGTH],
}

impl ScAddress {
    pub fn as_agent_id(&self) -> ScAgentID {
        ScAgentID::from_address(self)
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        address_to_bytes(self)
    }

    pub fn to_string(&self) -> String {
        address_to_string(self)
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn address_decode(dec: &mut WasmDecoder) -> ScAddress {
    let buf = dec.fixed_bytes(SC_ADDRESS_LENGTH);
    ScAddress {
        id: buf.try_into().expect("WTF?"),
    }
}

pub fn address_encode(enc: &mut WasmEncoder, value: &ScAddress) {
    enc.fixed_bytes(&value.id, SC_ADDRESS_LENGTH);
}

pub fn address_from_bytes(buf: &[u8]) -> ScAddress {
    let mut addr = ScAddress {
        id: [0; SC_ADDRESS_LENGTH],
    };
    let len = buf.len();
    if len == 0 {
        return addr;
    }

    // special case, ETH address has no type byte but different length
    if len == SC_LENGTH_ETH {
        addr.id[0] = SC_ADDRESS_ETH;
        addr.id[1..SC_LENGTH_ETH + 1].copy_from_slice(&buf[..SC_LENGTH_ETH]);
        return addr;
    }

    match buf[0] {
        SC_ADDRESS_ALIAS => {
            if len != SC_LENGTH_ALIAS {
                panic("invalid Address length: Alias");
            }
            addr.id[..SC_LENGTH_ALIAS].copy_from_slice(&buf[..SC_LENGTH_ALIAS]);
        }
        SC_ADDRESS_ED25519 => {
            if len != SC_LENGTH_ED25519 {
                panic("invalid Address length: Ed25519");
            }
            addr.id[..SC_LENGTH_ED25519].copy_from_slice(&buf[..SC_LENGTH_ED25519]);
        }
        SC_ADDRESS_NFT => {
            if len != SC_LENGTH_NFT {
                panic("invalid Address length: NFT");
            }
            addr.id[..SC_LENGTH_NFT].copy_from_slice(&buf[..SC_LENGTH_NFT]);
        }
        _ => panic("invalid Address type"),
    }
    addr
}

pub fn address_to_bytes(value: &ScAddress) -> Vec<u8> {
    match value.id[0] {
        SC_ADDRESS_ALIAS => {
            return value.id[..SC_LENGTH_ALIAS].to_vec();
        }
        SC_ADDRESS_ED25519 => {
            return value.id[..SC_LENGTH_ED25519].to_vec();
        }
        SC_ADDRESS_NFT => {
            return value.id[..SC_LENGTH_NFT].to_vec();
        }
        SC_ADDRESS_ETH => {
            return value.id[1..SC_LENGTH_ETH + 1].to_vec();
        }
        _ => panic("unexpected Address type"),
    }
    Vec::new()
}

pub fn address_from_string(value: &str) -> ScAddress {
    if !value.starts_with("0x") {
        return bech32_decode(value);
    }

    // ETH address, allow the common "0x0"
    if value == "0x0" {
        return address_from_bytes(&vec![0; SC_LENGTH_ETH]);
    }

    let bytes = hex_decode(value);
    if bytes.len() != SC_LENGTH_ETH {
        panic("invalid ETH address");
    }
    address_from_bytes(&bytes)
}

pub fn address_to_string(value: &ScAddress) -> String {
    if value.id[0] != SC_ADDRESS_ETH {
        return bech32_encode(value);
    }

    let mut hex = string_to_bytes(&hex_encode(&address_to_bytes(value)));
    let hash = hash_keccak(&hex[2..]).to_bytes();
    for i in 2..hex.len() {
        let mut hash_byte = hash[(i - 2) >> 1];
        if (i & 0x01) == 0 {
            hash_byte >>= 4;
        } else {
            hash_byte &= 0x0f;
        }
        if hex[i] > 0x39 && hash_byte > 7 {
            hex[i] -= 32;
        }
    }
    string_from_bytes(&hex)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableAddress {
    proxy: Proxy,
}

impl ScImmutableAddress {
    pub fn new(proxy: Proxy) -> ScImmutableAddress {
        ScImmutableAddress { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        address_to_string(&self.value())
    }

    pub fn value(&self) -> ScAddress {
        address_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable ScAddress in host container
pub struct ScMutableAddress {
    proxy: Proxy,
}

impl ScMutableAddress {
    pub fn new(proxy: Proxy) -> ScMutableAddress {
        ScMutableAddress { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: &ScAddress) {
        self.proxy.set(&address_to_bytes(&value));
    }

    pub fn to_string(&self) -> String {
        address_to_string(&self.value())
    }

    pub fn value(&self) -> ScAddress {
        address_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/scaddress.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from '../sandbox';
import {bech32Decode, bech32Encode, hexDecode, hexEncode, WasmDecoder, WasmEncoder, zeroes} from './codec';
import {Proxy} from './proxy';
import {bytesCompare} from './scbytes';
import {ScAgentID} from './scagentid';

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export const ScAddressAlias: u8 = 8;
export const ScAddressEd25519: u8 = 0;
export const ScAddressNFT: u8 = 16;
export const ScAddressEth: u8 = 32;

export const ScLengthAlias = 33;
export const ScLengthEd25519 = 33;
export const ScLengthNFT = 33;

export const ScAddressLength = ScLengthEd25519;
export const ScAddressEthLength = 21;

export class ScAddress {
    id: Uint8Array = zeroes(ScAddressLength);

    asAgentID(): ScAgentID {
        // agentID for address has Hname zero
        return ScAgentID.fromAddress(this);
    }

    public equals(other: ScAddress): bool {
        return bytesCompare(this.id, other.id) == 0;
    }

    // convert to byte array representation
    public toBytes(): Uint8Array {
        return addressToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        return addressToString(this);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// TODO address type-dependent encoding/decoding?
export function addressDecode(dec: WasmDecoder): ScAddress {
    const addr = new ScAddress();
    addr.id = dec.fixedBytes(ScAddressLength);
    return addr;
}

export function addressEncode(enc: WasmEncoder, value: ScAddress): void {
    enc.fixedBytes(value.id, ScAddressLength);
}

export function addressFromBytes(buf: Uint8Array | null): ScAddress {
    const addr = new ScAddress();
    if (buf === null || buf.length == 0) {
        return addr;
    }
    switch (buf[0]) {
        case ScAddressAlias:
            if (buf.length != ScLengthAlias) {
                panic('invalid Address length: Alias');
            }
            break;
        case ScAddressEd25519:
            if (buf.length != ScLengthEd25519) {
                panic('invalid Address length: Ed25519');
            }
            break;
        case ScAddressNFT:
            if (buf.length != ScLengthNFT) {
                panic('invalid Address length: NFT');
            }
            break;
        case ScAddressEth:
            if (buf.length != ScAddressEthLength) {
                panic('invalid Address length: Eth');
            }
            break;
        default:
            panic('invalid Address type');
    }
    for (let i = 0; i < buf.length; i++) {
        addr.id[i] = buf[i];
    }
    return addr;
}

export function addressToBytes(value: ScAddress): Uint8Array {
    switch (value.id[0]) {
        case ScAddressAlias:
            return value.id.slice(0, ScLengthAlias);
        case ScAddressEd25519:
            return value.id.slice(0, ScLengthEd25519);
        case ScAddressNFT:
            return value.id.slice(0, ScLengthNFT);
        case ScAddressEth:
            return value.id.slice(0, ScAddressEthLength);
        default:
            panic('unexpected Address type');
    }
    return new Uint8Array(0);
}

export function addressFromString(value: string): ScAddress {
    if (value.indexOf('0x') == 0) {
        const hexBytes = hexDecode(value);
        const b = new Uint8Array(hexBytes.length + 1);
        b[0] = ScAddressEth;
        b.set(hexBytes, 1);
        return addressFromBytes(b);
    }
    return bech32Decode(value);
}

export function addressToString(value: ScAddress): string {
    if (value.id[0] == ScAddressEth) {
        return hexEncode(value.id.slice(1, ScAddressEthLength));
    }
    return bech32Encode(value);
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableAddress {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return addressToString(this.value());
    }

    value(): ScAddress {
        return addressFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableAddress extends ScImmutableAddress {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScAddress): void {
        this.proxy.set(addressToBytes(value));
    }
}
